### ISSUE TITLE
[Docs]: Update AWS Provisioning & Deprovisioning Page

### ISSUE THREAD

**Issue by alechp** [konstructio/kubefirst-docs#568]
### What type of suggestions?

Missing Information

### What page is concerned?

- https://docs.kubefirst.io/aws/quick-start/install/cli
- https://docs.kubefirst.io/aws/deprovision/

### What version of the docs?

2.3.0+

### 1. Resolving 500 auth error that stems from using a normal IAM Admin role without AssumeAdmin
Also, debugging this and the 500 auth errors that came with it was a new addition in 2.3.0 (wasn't a problem in 2.2.17) and was a group effort from @CristhianF7 and @johndietz to fix it

Script that John sent:
```hcl
terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
      version = "4.67.0"
    }
  }
}
provider "aws" {
  region = "us-east-1"
}



resource "aws_iam_role" "assumed_admin" {
  name = "AssumedAdmin"
  ## this is 12 hours, both less secure and less annoying. default is 1 hr.
  max_session_duration = 43200
  ##
  ## note: below is a permissive role not intended for long term use. 
  ## it grants all iam users of the aws account the ability to assume the role AssumedAdmin
  ## which we create and attach aws role AdministratorAccess to. The value `:root` grants assume to the 
  ## whole account but you can replace root with your individual iam arn, or your role if appropriate.
  ## the value 111111111111 below should be replaced with your aws account id once you've
  ## understood the above implications. anyone with iam can assume the role while it's in 
  ## place like this. you can scope it down to your specific user, or across accounts, or whatever 
  ## you need. this assumes same account.
  ##


  assume_role_policy   = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "AllowAssumeRoleToIamUsers",
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::111my-aws-id111:root"
            },
            "Action": "sts:AssumeRole"
        }
    ]
}
EOF
}
resource "aws_iam_role_policy_attachment" "assumed_admin_admin_policy" {
  role       = aws_iam_role.assumed_admin.name
  policy_arn = "arn:aws:iam::aws:policy/AdministratorAccess"
}
```
User should `cat ~/.kubefirst` and ensure that IAM credentials match the newly generated STS

### 2. Using AWS STS for managing ephemeral tokens is a bit of a nuisance manually

I added this bash script as a convenience to help that along (might be worth including some variant of this):

```bash
function akauth() {
  unset AWS_ACCESS_KEY_ID
  unset AWS_SECRET_ACCESS_KEY
  unset AWS_SESSION_TOKEN

  local JSON=$(aws sts assume-role --role-arn "arn:aws:iam::111my-aws-account-id111:role/AssumedAdmin" --role-session-name "k-ray-eats-github-stars"
)
  local AWS_ACCESS_KEY_ID=$(echo $JSON | jq -r .Credentials.AccessKeyId)
  local AWS_SECRET_ACCESS_KEY=$(echo $JSON | jq -r .Credentials.SecretAccessKey)
  local AWS_SESSION_TOKEN=$(echo $JSON | jq -r .Credentials.SessionToken)

  local output="[default]\naws_access_key_id=$CS_AWS_ACCESS_KEY_ID\naws_secret_access_key=$CS_AWS_SECRET_ACCESS_KEY\n\n[kray]\naws_access_key_id=$AWS_ACCESS_KEY_ID\naws_secret_access_key=$AWS_SECRET_ACCESS_KEY\naws_session_token=$AWS_SESSION_TOKEN"

  mv ~/.aws/credentials ~/.aws/credentials.bak 
  echo $output >> ~/.aws/credentials
  export AWS_PROFILE="kray"
  echo $AWS_PROFILE
}
```

TLDR:
- Unset previously env vars
- Backup existing aws credentials file
- Automatically create a new ~/.aws/credentials which includes the previous default profile + the newly generated STS profile
- Update the AWS_PROFILE to point to the newly added profile in ~/.aws/credentials

This has to be run before the provisioner. For me that looks like:
```
akauth # run function above
./0101-run-k1-aws-via-brew.sh # run script that runs kubefirst aws create
```

For me that script (0101-run-k1-aws-via-brew.sh) looks like this:
```bash
# see what we're creating
export TARGET_CLOUD="aws"
echo "Creating a kubefirst instance on ${TARGET_CLOUD}"
echo "--------------------------------------------------"
echo "--gitops-template-url https://github.com/kubefirst/gitops-template"
echo "--gitops-template-branch main"
echo "--alerts-email $CS_CLUSTER_ADMIN"
echo "--domain-name $CS_CLUSTER_DOMAIN"
echo "--cluster-name $CS_CLUSTER_NAME"
echo "--github-org $CS_CLUSTER_GITHUB_ORG"
echo "--cloud-region $CS_CLUSTER_REGION"
echo "--dns-provider $CS_CLUSTER_DNS_PROVID"

# start creating
kubefirst ${TARGET_CLOUD} create \
  --gitops-template-url https://github.com/kubefirst/gitops-template \
  --gitops-template-branch main \
  --alerts-email $CS_CLUSTER_ADMIN \
  --domain-name $CS_CLUSTER_DOMAIN \
  --cluster-name $CS_CLUSTER_NAME \
  --github-org $CS_CLUSTER_GITHUB_ORG \
  --cloud-region $CS_CLUSTER_REGION \
  --dns-provider $CS_CLUSTER_DNS_PROVIDER
```

### 3. Add clarification that removing vault secrets won't work if the deploy never finished in the first place. This will result in a name conflict. Referring to this:
```bash
# update kubeconfig file locally so we can connect; assumes cluster creation went through in some capacity
aws eks update-kubeconfig --name $MY_CLUSTER_NAME --region $MY_CLUSTER_REGION

# requires kubeconfig to have been generated
export MY_VAULT_TOKEN=$(kubectl -n vault get secrets/vault-unseal-secret --template='{{index .data "root-token"}}' | base64 -d)

# requires vault root token
kubefirst terraform set-env \
  --vault-token $MY_VAULT_TOKEN \
  --vault-url https://vault.${MY_CLUSTER_DOMAIN} \
  --output-file .env
source .env
```

The CLI will return something like this:

<img width="1045" alt="Screenshot 2023-11-03 at 7 42 47 PM" src="https://github.com/kubefirst/docs/assets/3058939/9f0cd620-af3b-4676-89ea-d97ddfc38525">

Meanwhile, the logs from `kubefirst-kubefirst-api-*` pod will include some variant of `ResourceAlreadyExistsException`

For KMS, this is particularly finicky because you can't delete keys for 30 days. However, you can remove aliases: https://docs.aws.amazon.com/cli/latest/reference/kms/delete-alias.html (in CLI or console)
 
 ### 4. Failed to garbage collect required amount of images
 If user sees this while running k8s via docker desktop, there's a good chance that they've accumulated a bunch of volumes. I made this mistake because I have scripts to clear containers & images, but didn't touch volumes
 
![image](https://github.com/kubefirst/docs/assets/3058939/dc23591a-09b4-47e8-b87f-afafe1e49a01)
![image](https://github.com/kubefirst/docs/assets/3058939/d7b88a21-b115-4b27-bb14-2a01fd9811bc)

Also, might make sense to include a sane default for local memory requirements on the provisioning ends
<img width="682" alt="Screenshot 2023-11-03 at 8 15 28 PM" src="https://github.com/kubefirst/docs/assets/3058939/ab2ab59c-8eeb-47df-96be-ab7f86156bbd">

### 5. While deprovisioning, there are multiple things to consider
In addition to nuking the remote and local clusters, it's helpful to clean up completely to make sure there are no scale artifacts that will screw up the next deploy. This includes local kubefirst config, k1 state directory, github repos and github teams

All in all, this is the script I run each time:
```bash
# remove kubefirst artifacts
rm ~/.kubefirst
rm -Rf ~/.k1

# remove github repos
gh repo delete ${CS_CLUSTER_GITHUB_ORG}/gitops --yes
gh repo delete ${CS_CLUSTER_GITHUB_ORG}/metaphor --yes

# delete teams
export TEAM_ADMINS="admins"
export TEAM_DEVS="developers"
echo "Deleting teams ${TEAM_ADMINS} & ${TEAM_DEVS} "
gh api \
  --method DELETE \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  /orgs/${CS_CLUSTER_GITHUB_ORG}/teams/${TEAM_ADMINS}

gh api \
  --method DELETE \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  /orgs/$CS_CLUSTER_GITHUB_ORG/teams/${TEAM_DEVS}
  ``` 
### 6. The aws provision page references the `vultr` (not aws) provider
https://docs.kubefirst.io/aws/quick-start/install/cli
<img width="603" alt="Screenshot 2023-11-03 at 8 06 51 PM" src="https://github.com/kubefirst/docs/assets/3058939/f48b91b9-0543-4963-ad33-0d76ca8b4a99">

### 7. If a cluster takes too long to provision the STS token will expire. This means that trying to run terraform init/terraform destroy will fail

It'll throw an error like:
```
error configuring S3 Backend: error validating provider credentials: error calling sts:GetCallerIdentity: ExpiredToken: The security token included in the request is expired
status code: 403
```

In this instance the user just needs to reset to their default IAM credential with admin privilege, example:
```
export AWS_PROFILE=default
```

Then run `aws configure` to ensure that it's showing the last 4 characters of the correct IAM. After that, user should be able to terraform init/destroy properly

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by alechp** [comment]
This issue is regarding on ongoing issue that @CristhianF7 has been helping me debug
 <img width="1191" alt="Screenshot 2023-11-03 at 8 03 25 PM" src="https://github.com/kubefirst/docs/assets/3058939/b8147f79-7553-4a49-bc16-0845933b61ec">

Link to relevant slack thread: https://kubefirst.slack.com/archives/C03U34WJ7FW/p1698957656303509

**Comment by fharper** [comment]
Thanks @alechp! There are a lot to unpack here, I'll check that once I'm back from KubeCon :)

**Comment by alechp** [comment]
Re: the STS ephemeral creds: 

https://github.com/kubefirst/kubefirst/issues/1909

<img width="1264" alt="Screenshot 2023-11-14 at 2 47 31 PM" src="https://github.com/kubefirst/docs/assets/3058939/0f66c7ca-4e29-49c5-8470-3b6f5d29c55c">

<img width="1266" alt="Screenshot 2023-11-14 at 2 47 42 PM" src="https://github.com/kubefirst/docs/assets/3058939/f1907c0e-3287-4234-9eaa-d9b8cb893907">

--- 

1. Launch terminal
2. Run `akauth` (or equivalent function) to connect to the cluster:
```sh
function akauth() {
  unset AWS_ACCESS_KEY_ID
  unset AWS_SECRET_ACCESS_KEY
  unset AWS_SESSION_TOKEN

  local JSON=$(aws sts assume-role --role-arn "arn:aws:iam::111my-aws-account-id111:role/AssumedAdmin" --role-session-name "k-ray-eats-github-stars"
)
  local AWS_ACCESS_KEY_ID=$(echo $JSON | jq -r .Credentials.AccessKeyId)
  local AWS_SECRET_ACCESS_KEY=$(echo $JSON | jq -r .Credentials.SecretAccessKey)
  local AWS_SESSION_TOKEN=$(echo $JSON | jq -r .Credentials.SessionToken)

  local output="[default]\naws_access_key_id=$CS_AWS_ACCESS_KEY_ID\naws_secret_access_key=$CS_AWS_SECRET_ACCESS_KEY\n\n[kray]\naws_access_key_id=$AWS_ACCESS_KEY_ID\naws_secret_access_key=$AWS_SECRET_ACCESS_KEY\naws_session_token=$AWS_SESSION_TOKEN"

  mv ~/.aws/credentials ~/.aws/credentials.bak 
  echo $output >> ~/.aws/credentials
  export AWS_PROFILE="kray"
  echo $AWS_PROFILE
}
```
3. Run `kubectl get cm -n kube-system aws-auth -oyaml`
4. Open a new terminal
5. In the new terminal, echo $AWS_PROFILE and make sure that it shows your default IAM credential (ie. the one you normally use for AWS, not the ephemeral role generated by `aws sts assume-role...`)
6. Grab the ARN value for this IAM. Convenient script to get just the relevant value, print it out and add to your clipboard:
```sh
export AWSGCI=$(aws sts get-caller-identity | jq -r .Arn)
echo $AWSGCI
echo $AWSGCI | pbcopy
```
7. Update the aws-auth.yaml to something like:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: kube-system
data:
  mapRoles: |
    - groups:
      - system:bootstrappers
      - system:nodes
      rolearn: arn:aws:iam::111111111111:role/default_node_group-eks-node-group-20231113155100128100000002
      username: system:node:{{EC2PrivateDNSName}}
    - groups:
      - system:masters
      rolearn: arn:aws:iam::111111111111:user/alechp
      username: admin
```
8. Add this to your gitops repo, example:
```sh
 cp aws-auth.yaml ../repos/gitops/registry/clusters/cscloud
```
9. Github action should pick it up and apply like so:
> <img width="367" alt="Screenshot 2023-11-14 at 2 54 44 PM" src="https://github.com/kubefirst/docs/assets/3058939/dcb36ba5-cfa3-4f23-96d8-7e0fc71ed118">
10. After confirming it updates, open a new terminal using your admin IAM user (run `echo $AWS_PROFILE` to make sure it's what you expect... ie. should be the one *you* created... NOT the AWS STS generated ephemeral credentials)
11. Generate your kubeconfig
```sh
aws eks update-kubeconfig --region $MY_CLUSTER_REGION --name $MY_CLUSTER_NAME
```
12. After all said and done, you should be able to access the cluster with your default admin credentials in ways you're accustomed (eg. Lens SS):
<img width="934" alt="Screenshot 2023-11-15 at 2 19 45 PM" src="https://github.com/kubefirst/docs/assets/3058939/42e84b84-14d1-43e5-bb96-55856f3e63b0">

**Comment by fharper** [comment]
@alechp I came back from KubeCon, and got swamped by other stuff, but I'm finally taking the time to look at this issue, sorry for the delay!

**Comment by fharper** [comment]
As a side note, it may take a bit more time on your side, but it will make it easier to follow up, and fix issues if you create one issue per specific topic, problem, or suggestion :)

With that said, let me get back to you one by one (as soon as I can get back about specific ones) so it's easier for you, and me to manage the discussions.

**Comment by fharper** [comment]
> 4. Failed to garbage collect required amount of images

I do not have this issue so I assume it may have happened when `destroy` was the way to delete cluster created by the CLI directly and you had issues, or when `launch down` wasn't possible. Now that you cleaned everything, let me know if it's happening again in a new issue please.

**Comment by fharper** [comment]
> Also, might make sense to include a sane default for local memory requirements on the provisioning ends

So we had this in the k3d documentations https://docs.kubefirst.io/k3d/explore/limitations#docker-desktop-recommendation but I added information about virtual disk + swap minimal requirements. Your values are great, but someone can run `launch up` or k3d with less. I also moved the content in a better place, and on all clouds.

See https://github.com/kubefirst/docs/pull/585

**Comment by fharper** [comment]
> 3. Add clarification that removing vault secrets won't work if the deploy never finished in the first place. This will result in a name conflict.

Good call! I added more information about deprovisioning if Vault didn't deployed correctly, and the `kubefirst terraform set-env` step failed: https://github.com/kubefirst/docs/pull/586

**Comment by fharper** [comment]
> 6. The aws provision page references the vultr (not aws) provider

Good eyes, it was an issue in all cloud pages, not just AWS. It's fixed in https://github.com/kubefirst/docs/pull/587

**Comment by fharper** [comment]
> 5. While deprovisioning, there are multiple things to consider
> This includes local kubefirst config, k1 state directory

This is done with the last step of deprovisioning https://docs.kubefirst.io/civo/deprovision

![CleanShot 2023-12-06 at 11 02 15@2x](https://github.com/kubefirst/docs/assets/1170989/36d80c75-6e57-4030-9b59-137cdb28ee58)

I'll added this steps on it's own section (header) and made it clearer it's not optional even if you use the console to create your cluster https://github.com/kubefirst/docs/pull/588

> github repos and github teams

Those are done with the Git Terraform step https://docs.kubefirst.io/civo/deprovision#git .

Let me know if those aren't working for you, we'll need to investigate as it will be a bug.

**Comment by fharper** [comment]
> Re: the STS ephemeral creds:

Thanks for all the steps, but I won't document this since we have the issue https://github.com/kubefirst/kubefirst/issues/1909 created, and once it's fixed, I don't think we would have to do that again.

**Comment by fharper** [comment]
And finally...

> 1. Resolving 500 auth error that stems from using a normal IAM Admin role without AssumeAdmin

I added some information on assumed admin role at https://github.com/kubefirst/kubefirst/pull/1946

> 2. Using AWS STS for managing ephemeral tokens is a bit of a nuisance manually

I added our own script we use internally, which is kinda like yours. https://github.com/kubefirst/kubefirst/blob/main/tools/aws-assume-role.sh

It's also linked in the docs.

> Re: the STS ephemeral creds:

I added John's Terraform plan which is also linked in the docs https://github.com/kubefirst/kubefirst/blob/main/tools/aws-create-role.tf

**Comment by fharper** [comment]
With the previous comment, I think I address all suggestions!

I'll close the issue for my own sake!

If there is anything else, feel free to reopen, or create another issue.

Thanks a lot for all the suggestions, they were all amazing!
